index.js

```diff
@@ -0,0 +1,254 @@
+            module.addChunk(targetChunk);  // modules can appear in multiple chunks
+        }
+
+        // Add chunk dependencies.
+        targetChunk.parents = [...chunkDeps];
+        for (const chunk of chunkDeps) {
+            chunk.addChunk(targetChunk);
+        }
+
+        // It's possible that moduleFilter function filters out all modules
+        // from a particular chunk in which case it isn't actually affected.
+        // TODO(kevinb): figure out what to do when extracting common
+        // chunks from async chunks, see CommonsChunkPlugin
+        const actuallyAffectedChunks = new Set();
+        for (const module of commonModules) {
+            for (const chunk of affectedChunks) {
+                if (module.removeChunk(chunk)) {
+                    actuallyAffectedChunks.add(chunk);
+                }
+            }
+        }
+
+        // Copied from makeTargetChunkParentOfAffectedChunks in
+        // CommonsChunkPlugin.js
+        for (const chunk of actuallyAffectedChunks) {
+            chunk.parents = [...chunk.parents, targetChunk];
+            targetChunk.addChunk(chunk);
+
+            for (const entrypoint of chunk.entrypoints) {
+                entrypoint.insertChunk(targetChunk, chunk);
+            }
+        }
+    }
+
+    // Return module and chunk dependencies for the given commonModules.
+    // Module dependencies are determined recursively.  Whenever a module
+    // appearing in globalModulesSet is encountered, recursion is terminated
+    // and any chunks that the module appears in are return as chunk
+    // dependencies.
+    getDependencies(commonModules, globalModulesSet) {
+        const moduleDeps = new Set();
+        const chunkDeps = new Set();
+
+        const walkDeps = (module) => {
+            for (const dep of module.dependencies) {
+                if (dep.module && dep.module.resource) {
+                    // ignore dependencies
+                    if (!globalModulesSet.has(dep.module)) {
+                        moduleDeps.add(dep.module);
+                        walkDeps(dep.module);
+                    } else {
+                        for (const chunk of dep.module.chunks) {
+                            // Don't add entry chunks
+                            // TODO(kevinb): remove dependencies first before
+                            // calculating chunk dependencies, we'll have to
+                            // split this method in two so that we can grab
+                            // all module deps first, remove them, then compute
+                            // the chunk deps.
+                            // TODO(kevinb): write a failing test case without
+                            // this check, also, this check is specific to
+                            // webapp's structure, we need to avoid that.
+                            if (!chunk.name.startsWith('javascript/')) {
+                                chunkDeps.add(chunk);
```

Comment:
Without this check, all of the entry chunks were being added as dependencies of vendor and shared because the files that they shared with those modules hadn't been removed yet.

Reply:
Hmm... ok! Seems like this should probably be a STOPSHIP if we're releasing this as an independent lib.

Reply:
I definitely want to fix this before landing this diff, let alone releasing this as an independent library.